### PR TITLE
Data augmentations

### PR DESCRIPTION
**Pull request by bernardohenz**
Data augmentation on feature-space and on spectrogram (using tf operations).
Check [here](https://colab.research.google.com/drive/1MolmJnPWzo1JbU6uA6dHPhnbHsPdzesA) to see how each augmentation change the spectrogram.

Here is a benchmark I've run *on our training set*, training for 40 epochs (no early-stop), comparing the augmentations.

|                        |                                 |            |           |            |               |
|------------------------|---------------------------------|------------|-----------|------------|---------------|
| Training for 40 epochs |                                 |            |           |            |               |
| Experiment                   | Model                           | WER (best) | Test loss | best epoch | training time |
| exp_2                 | Baseline (700 hidden)           | 0.04617    | 18.232    | 39         | 8:49:43       |
| exp_3                 | Baseline+Layernorm              | 0.04569    | 17.902    | 37         | 8:51:00       |
| exp_4                 | Baseline+Layernorm+MaskingAug   | 0.04331    | 17.568    | 30         | 9:10:54       |
|  **exp_5**                 | exp_4+dropoutAug (0.95)         | 0.04422    | 17.135    | 39         | 9:45:52       |
| exp_6                 | exp_4+dense_time_warp          | 0.04481    | 17.898    | 33         | 25:45:18***      |
| **exp_7**                 | exp_4+speedup (std=0.05)        | 0.04482    | 17.447    | 35         | 9:24:54       |
| exp_8                 | exp_4+tempo&pitch              | 0.04929    | 18.264    | 39         | 9:33:37       |
| **exp_9**                 | exp_4+additive_noise (std=0.05) | 0.04445    | 17.363    | 36         | 9:17:57       |
| exp_10                | exp_4+multiplicative_noise (std=0.05)     | 0.04484    | 17.756    | 38         | 9:21:41       |
| **task_11**                | exp_4+**bolds**                   | **0.04226**    | **17.155**    | 36         | 9:53:24       |

As you can see, the augmentations that seem improve test-loss is the **masking freq-time**, **dropoutAug**, **speedUp**, and **additive_noise**. Nonetheless, I've included the other augs in the PR (maybe you can find better parameters, or make it work).

IMO, augmentations based on image warping are **incredibly slow**, being the first to be excluded. The other augmentations seem pretty fast, as they are implemented using tf operations.

PS: my layer-norm implementation is slightly different from the one on yours branch, dunno which one is the best.

### CODE REVIEW

**Comment by cahuja1992** [comment]
@bernardohenz This is nice work. I have few doubts.
1. When you are trying to integrate with [feeding.py](https://github.com/bernardohenz/DeepSpeech/blob/ca3a164615956baacb8228e277d52dc01ae1ff28/util/feeding.py#L35), this line actually calculate the spectrogram that is equivalent to tf.stft. Bu the spec augment methods like`augment_sparse_deform` needs log mel-spectrogram. Therefore, I doubt whether this is correct or not. Or else log-mel-spectrogram needs to be calculated 
while using something like this [log_mel_spectrograms](https://www.tensorflow.org/api_docs/python/tf/signal/mfccs_from_log_mel_spectrograms)

2. Method spsi to convert Spectrogram to wav, does works with librosa but have you tried it with tensorflow contrib_audio. audio_spectrogram ?

**Comment by bernardohenz** [comment]
@cahuja1992 I've added a comparison of TF and librosa spectrogram computation ([link](https://colab.research.google.com/drive/1MolmJnPWzo1JbU6uA6dHPhnbHsPdzesA)), and it seems that both work similarly, i.e., it is possible to (partially) reconstruct the audio from its spectrogram (either if it was computed with librosa or TF).

With respect to log-space, I don't believe this makes any difference in practice. All the augs either deform or rescale the spectrogram, which implies on interpolating values. Although interpolating on linear or log-space makes difference, I think the space of interpolation is not relevant in practice. Concluding: performing the augmentations on linear or log space will result in almost-like spectrograms.,

**Comment by cahuja1992** [comment]
> @cahuja1992 I've added a comparison of TF and librosa spectrogram computation ([link](https://colab.research.google.com/drive/1MolmJnPWzo1JbU6uA6dHPhnbHsPdzesA)), and it seems that both work similarly, i.e., it is possible to (partially) reconstruct the audio from its spectrogram (either if it was computed with librosa or TF).
> 
> With respect to log-space, I don't believe this makes any difference in practice. All the augs either deform or rescale the spectrogram, which implies on interpolating values. Although interpolating on linear or log-space makes difference, I think the space of interpolation is not relevant in practice. Concluding: performing the augmentations on linear or log space will result in almost-like spectrograms.,

Thanks. Its just in the SpecAugment paper, google mentioned that they augment over log melspectoram.

**Comment by cahuja1992** [comment]
@bernardohenz :
When exp_2(2048 n_hidden), exp_3 and exp_4 is tried on Common Voice Dataset. But could not reproduce the desired behavior. Could you please share which dataset you used and also the parameters?

**Comment by bernardohenz** [comment]
@cahuja1992 The benchmark was run on our private pt-br dataset. Unfortunately, we can't share right now. I couldn't benchmark on the Common Voice Dataset because all our machines are busy.

Thus, you will probably get different WER and loss values, but I believe that the augmentations should improve your numbers similarly as they improved ours.

**Comment by khsinclair** [comment]
We've been using similar augmentation to train both Mozilla DeepSpeech and Baidu DeepSpeech 2, and have learned a few things.  Our task is quite different: we're retraining the 0.5.1 model with 2000 hours of noisy conversational English, with a baseline WER around 50%.  With Mozilla DeepSpeech, masking in the MFCC domain seems to work better than masking the spectrograms.

The DeepSpeech 2 network input is normalized log-valued linear-frequency spectrograms, and we use essentially the SpecAugment masking policies, and we see the same effect described in the google paper of higher training loss producing lower validation loss, postponing overtraining.  Accuracy on 4 and 15 hour test sets improves from 31% to 29% WER with this augmentation.

Mozilla DeepSpeech network input is non-normalized MFCC coefficients, which I believe includes c0.  We tried a number of policies applying SpecAugment masking to the mel-frequency magnitude or log-valued spectrograms, before the MFCC transform, and never got any consistent accuracy gain.  We tried masking with zeros, with mean values, with noise energy at various SNRs.

What does seem to help is augmentation in the MFCC domain.  Here, we just randomly truncate the input, setting higher-order coefficients to zero.  So the SpecAugment "F" mask is interpreted as "zero every coefficient above random(8,26) for every frame in the utterance".  "T" mask just sets to zero, but aggressive T masking policy doesn't work as well as in the Baidu system.

With this MFCC-domain masking augmentation, accuracy improves from 37.5% to 36% WER.  That's actually good enough for now, though we want to do better next year.

It's not clear to me why DeepSpeech is using MFCC input, I would think normalized log-valued mel-frequency spectrograms would provide better information.  Maybe we'll try that sometime.

**Comment by bernardohenz** [comment]
@khsinclair thanks for the tips. I'll try to implement and evaluate this masking on the MFCC-features domain and post it here ASAP.

**Comment by rhamnett** [comment]
Hi, how does https://github.com/mozilla/DeepSpeech/pull/2252 impact this?

**Comment by bernardohenz** [comment]
@rhamnett All augmentations of this PR works over spectrogram and/or feature space. All of them are implemented using tensorflow funcs. Thus I believe this should be considerably faster compared to augmenting audio data.

**Comment by cahuja1992** [comment]
@bernardohenz 
In exp_3 you mentioned you used Layernorm, on all the layers you did layernorm?

**Comment by bernardohenz** [comment]
@cahuja1992 after all dense layers except the last one. You can replace the code on https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L65 by the following:
```python
def dense(name, x, units, dropout_rate=None, relu=True, layer_norm=False):
    with tf.variable_scope(name):
        bias = variable_on_cpu('bias', [units], tf.zeros_initializer())
        weights = variable_on_cpu('weights', [x.shape[-1], units], tf.contrib.layers.xavier_initializer())

    output = tf.nn.bias_add(tf.matmul(x, weights), bias)

    if relu:
        output = tf.minimum(tf.nn.relu(output), FLAGS.relu_clip)

    if layer_norm:
        output = tf.contrib.layers.layer_norm(output)

    if dropout_rate is not None:
        output = tf.nn.dropout(output, rate=dropout_rate)

    return output
```

**Important: If you train using the layernorm, you must remember to evaluate/export using the layernorm as well.

**Comment by cahuja1992** [comment]
> @cahuja1992 after all dense layers except the last one. You can replace the code on https://github.com/mozilla/DeepSpeech/blob/master/DeepSpeech.py#L65 by the following:
> 
> ```python
> def dense(name, x, units, dropout_rate=None, relu=True, layer_norm=False):
>     with tf.variable_scope(name):
>         bias = variable_on_cpu('bias', [units], tf.zeros_initializer())
>         weights = variable_on_cpu('weights', [x.shape[-1], units], tf.contrib.layers.xavier_initializer())
> 
>     output = tf.nn.bias_add(tf.matmul(x, weights), bias)
> 
>     if relu:
>         output = tf.minimum(tf.nn.relu(output), FLAGS.relu_clip)
> 
>     if layer_norm:
>         output = tf.contrib.layers.layer_norm(output)
> 
>     if dropout_rate is not None:
>         output = tf.nn.dropout(output, rate=dropout_rate)
> 
>     return output
> ```
> 
> **Important: If you train using the layernorm, you must remember to evaluate/export using the layernorm as well.

Thanks @bernardohenz , you didn't use layer norm for the LSTM layer ?

File: util/sparse_image_warp.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    boundary_point_flows = _expand_to_minibatch(boundary_point_flows, batch_size)
+
+    merged_control_point_locations = array_ops.concat([control_point_locations, boundary_point_locations], 1)
+
+    merged_control_point_flows = array_ops.concat([control_point_flows, boundary_point_flows], 1)
+
+    return merged_control_point_locations, merged_control_point_flows
+
+
+def sparse_image_warp(image,
```
Comment (reuben) [inline util/sparse_image_warp.py]: Why inline this into the repo instead of using `tf.contrib.image.sparse_image_warp`? Was it modified?

Comment (bernardohenz) [reply util/sparse_image_warp.py]: This is from where I got the code: https://github.com/tensorflow/tensorflow/pull/30904

**Comment by reuben** [comment]
I'm testing this code against our Mandarin data to see how it performs against a baseline. So far it looks promising performance wise, matching your results of very little performance overhead. I think we should merge this, but remove the sparse_image_warp transform and code since it's too slow to be useful.

**Comment by reuben** [comment]
Merged in #2352!
